clients/java/src/main/java/io/camunda/zeebe/client/impl/command/CorrelateMessageCommandImpl.java

```diff
@@ -0,0 +1,91 @@
+import io.camunda.zeebe.client.api.command.CorrelateMessageCommandStep1.CorrelateMessageCommandStep2;
+import io.camunda.zeebe.client.api.command.CorrelateMessageCommandStep1.CorrelateMessageCommandStep3;
+import io.camunda.zeebe.client.api.command.FinalCommandStep;
+import io.camunda.zeebe.client.api.response.CorrelateMessageResponse;
+import io.camunda.zeebe.client.impl.http.HttpClient;
+import io.camunda.zeebe.client.impl.http.HttpZeebeFuture;
+import io.camunda.zeebe.client.impl.response.CorrelateMessageResponseImpl;
+import io.camunda.zeebe.client.protocol.rest.MessageCorrelationRequest;
+import io.camunda.zeebe.client.protocol.rest.MessageCorrelationResponse;
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import org.apache.hc.client5.http.config.RequestConfig;
+
+public class CorrelateMessageCommandImpl extends CommandWithVariables<CorrelateMessageCommandImpl>
+    implements CorrelateMessageCommandStep1,
+        CorrelateMessageCommandStep2,
+        CorrelateMessageCommandStep3 {
+
+  private final HttpClient httpClient;
+  private final JsonMapper jsonMapper;
+  private final MessageCorrelationRequest request = new MessageCorrelationRequest();
+  private final RequestConfig.Builder httpRequestConfig;
+
+  public CorrelateMessageCommandImpl(final HttpClient httpClient, final JsonMapper jsonMapper) {
+    super(jsonMapper);
+    this.httpClient = httpClient;
+    this.jsonMapper = jsonMapper;
+    httpRequestConfig = httpClient.newRequestConfig();
+  }
+
+  @Override
+  public CorrelateMessageCommandStep2 messageName(final String messageName) {
+    request.setName(messageName);
+    return this;
+  }
+
+  @Override
+  public CorrelateMessageCommandStep3 tenantId(final String tenantId) {
+    request.setTenantId(tenantId);
+    return this;
+  }
+
+  @Override
+  public CorrelateMessageCommandStep3 correlationKey(final String correlationKey) {
+    request.setCorrelationKey(correlationKey);
+    return this;
+  }
+
+  @Override
+  public CorrelateMessageCommandStep3 withoutCorrelationKey() {
+    return this;
+  }
+
+  @Override
+  public FinalCommandStep<CorrelateMessageResponse> requestTimeout(final Duration requestTimeout) {
+    httpRequestConfig.setResponseTimeout(requestTimeout.toMillis(), TimeUnit.MILLISECONDS);
+    return this;
+  }
+
+  @Override
+  public ZeebeFuture<CorrelateMessageResponse> send() {
+    final HttpZeebeFuture<CorrelateMessageResponse> result = new HttpZeebeFuture<>();
+    final CorrelateMessageResponseImpl response = new CorrelateMessageResponseImpl(jsonMapper);
+    httpClient.post(
+        "/messages/correlation",
+        jsonMapper.toJson(request),
+        httpRequestConfig.build(),
+        MessageCorrelationResponse.class,
+        response::setResponse,
+        result);
+    return result;
+  }
+
+  @Override
+  protected CorrelateMessageCommandImpl setVariablesInternal(final String variables) {
+    request.setVariables(jsonMapper.fromJsonAsMap(variables));
```

Comment:
This methods gets called from the interface which actually exposes the variables method for users. This method is only internal. These methods already do a JSON validation, so it's not neccessary to them here. E.g.:

```java
   public T variables(final String variables) {
    ArgumentUtil.ensureNotNull("variables", variables);
    return setVariablesInternal(objectMapper.validateJson("variables", variables));
  }
```

Reply:
Sorry I gave you a wrong explanation, if you provide a VALID json string, but NOT VALID for being transformed into a `Map` the command fails with the following error:

```
io.camunda.zeebe.client.api.command.InternalClientException: Failed to deserialize json '[]' to 'Map<String, Object>'
```

Reply:
The gRPC behaviour in those cases is different, because is sending the request without problems, but the engine then send back an exception when trying to wrap the string variable into a document ([here](https://github.com/camunda/camunda/blob/main/zeebe/msgpack-value/src/main/java/io/camunda/zeebe/msgpack/value/DocumentValue.java))

Reply:
That seems valid no? Variables must be in a map-like structure. We actively create this exception ourselves in the `ZeebeObjectMapper` 😄 

```java
  @Override
  public Map<String, Object> fromJsonAsMap(final String json) {
    try {
      return objectMapper.readValue(json, MAP_TYPE_REFERENCE);
    } catch (final IOException e) {
      throw new InternalClientException(
          String.format("Failed to deserialize json '%s' to 'Map<String, Object>'", json), e);
    }
  }
```

Reply:
🤔 Hmm we do accept `[]` in message publish. I'll have a look.

Reply:
Sure I can add it. Thanks for raising!

Reply:
@nicpuppa All set, pleased approve when you approve 😄